OvmfPkg/Library/UefiApplicationOverride/ApplicationOverride.c

```diff
@@ -0,0 +1,129 @@
+/** @file
+  Entry point library instance to a UEFI application. This version is specific to QEMU, and ties
+  dispatch of the application in question on the value of a QEMU fw_cfg boolean
+  variable which is referenced by name via a fixed pointer PCD.
+
+  Copyright (c) 2024, Red Hat, Inc.
+
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+**/
+
+#include <Uefi.h>
+#include <Library/UefiApplicationEntryPoint.h>
+#include <Library/BaseLib.h>
+#include <Library/UefiBootServicesTableLib.h>
+
+#include <Library/QemuFwCfgSimpleParserLib.h>
+
+/**
+  Entry point to UEFI Application.
+
+  This function is the entry point for a UEFI Application. This function must call
+  ProcessLibraryConstructorList(), ProcessModuleEntryPointList(), and ProcessLibraryDestructorList().
+  The return value from ProcessModuleEntryPointList() is returned.
+  If _gUefiDriverRevision is not zero and SystemTable->Hdr.Revision is less than _gUefiDriverRevison,
+  then return EFI_INCOMPATIBLE_VERSION.
+
+  @param  ImageHandle                The image handle of the UEFI Application.
+  @param  SystemTable                A pointer to the EFI System Table.
+
+  @retval  EFI_SUCCESS               The UEFI Application exited normally.
+  @retval  EFI_INCOMPATIBLE_VERSION  _gUefiDriverRevision is greater than SystemTable->Hdr.Revision.
+  @retval  Other                     Return value from ProcessModuleEntryPointList().
+
+**/
+EFI_STATUS
+EFIAPI
+_ModuleEntryPoint (
+  IN EFI_HANDLE        ImageHandle,
+  IN EFI_SYSTEM_TABLE  *SystemTable
+  )
+{
+  EFI_STATUS     Status;
+  BOOLEAN        Enabled;
+  RETURN_STATUS  RetStatus;
+
+  if (_gUefiDriverRevision != 0) {
+    //
+    // Make sure that the EFI/UEFI spec revision of the platform is >= EFI/UEFI spec revision of the application.
+    //
+    if (SystemTable->Hdr.Revision < _gUefiDriverRevision) {
+      return EFI_INCOMPATIBLE_VERSION;
+    }
+  }
+
+  //
+  // Call constructor for all libraries.
+  //
+  ProcessLibraryConstructorList (ImageHandle, SystemTable);
+
+  RetStatus = QemuFwCfgParseBool (
+                FixedPcdGetPtr (PcdEntryPointOverrideFwCfgVarName),
+                &Enabled
+                );
+
+  if (!RETURN_ERROR (RetStatus) && !Enabled) {
+    //
+    // The QEMU fw_cfg variable tells us not to load this image.  So abort.
+    //
+    Status = EFI_ABORTED;

```

Comment:
Probably helpful to print a message to the screen here, so the user knows why the app does not start.
```
    Print (L"...\n");
    gBS->Stall (2000000);   // two secs delay
```

Reply:
Good idea, added in the new version!